python-sdk/actionml/__init__.py

```diff
@@ -0,0 +1,634 @@
+        :param event_id: event id returned by the EventServer when creating the event.
+        :returns: AsyncRequest object.
+        """
+        request = AsyncRequest("GET", self._add_segment(event_id))
+        request.set_response_handler(self._ok_response_handler)
+        self._connection.make_request(request)
+        return request
+
+    def get(self, event_id):
+        """Synchronouly get an event from PIO Kappa Server."""
+        return self.async_get(event_id).get_response()
+
+    def async_delete(self, event_id):
+        """Asynchronouly delete an event from PIO Kappa Server.
+    :param event_id: event id returned by the EventServer when creating the
+      event.
+    :returns:
+      AsyncRequest object.
+    """
+        request = AsyncRequest("DELETE", self._add_segment(event_id))
+        request.set_response_handler(self._delete_response_handler)
+        self._connection.make_request(request)
+        return request
+
+    def delete(self, event_id):
+        """Synchronously delete an event from PIO Kappa Server."""
+        return self.async_delete(event_id).get_response()
+
+    ## Below are helper functions
+
+    def async_set_user(self, uid, properties=None, event_time=None):
+        """
+        Set properties of a user.
+        Wrapper of async_create function, setting event to "$set" and entity_type
+        to "user".
+        """
+        if properties is None:
+            properties = {}
+        return self.async_create(
+            event="$set",
+            entity_type="user",
+            entity_id=uid,
+            properties=properties,
+            event_time=event_time,
+        )
+
+    def set_user(self, uid, properties=None, event_time=None):
+        """Set properties of a user"""
+        if properties is None:
+            properties = {}
+        return self.async_set_user(uid, properties, event_time).get_response()
+
+    def async_unset_user(self, uid, properties, event_time=None):
+        """
+        Unset properties of an user.
+        Wrapper of async_create function, setting event to "$unset" and entity_type
+        to "user".
+        """
+        # check properties={}, it cannot be empty
+        return self.async_create(
+            event="$unset",
+            entity_type="user",
+            entity_id=uid,
+            properties=properties,
+            event_time=event_time,
+        )
+
+    def unset_user(self, uid, properties, event_time=None):
+        """Set properties of a user"""
+        return self.async_unset_user(uid, properties, event_time).get_response()
+
+    def async_delete_user(self, uid, event_time=None):
+        """
+        Delete a user.
+        Wrapper of async_create function, setting event to "$delete" and entity_type
+        to "user".
+        """
+        return self.async_create(
+            event="$delete",
+            entity_type="user",
+            entity_id=uid,
+            event_time=event_time)
+
```

Comment:
Stick to REST defined API, we will never have a "user delete" because notion of a user is Template specific. Only resources defined in REST can be manipulated and the admin CLI only accesses these, dataset, engine, command.

These are the same rules as the Java SDK with the addition of the admin CLI + admin REST API. For instance queries are json strings, the structure is Template specific and not defined in the Python SDK or Python CLI.